src/KeyringController.ts

```diff
@@ -903,12 +919,27 @@ class KeyringController extends EventEmitter {
         throw new TypeError(KeyringControllerError.WrongPasswordType);
       }
 
-      vault = await this.encryptor.decrypt(password, encryptedVault);
+      vault = await this.#encryptor.decrypt(password, encryptedVault);
       this.password = password;
     }
 
+    if (!isSerializedKeyringsArray(vault)) {
+      throw new Error(KeyringControllerError.VaultDataError);
+    }
+
     await Promise.all(vault.map(this.#restoreKeyring.bind(this)));
     await this.updateMemStoreKeyrings();
+
+    if (
+      this.password &&
+      this.#encryptor.updateVault &&
+      (await this.#encryptor.updateVault(encryptedVault, this.password)) !==
```

Comment:
Yes, that's quite bad. The best way would be to expose `isVaultUpdated` from `browser-passworder`, because I don't think there's a nice way to inline that logic here, as `eth-keyring-controller` would have to know the latest vault parameters (`DEFAULT_DERIVATION_PARAMS` in `browser-passworder`, which is also not exposed)